tests/python/test_linear.py

```diff
@@ -1,87 +1,49 @@
-                          atol=tol).all(), (weights, enet.coef_)
-        assert np.isclose(enet_pred, pred, rtol=tol, atol=tol).all(), (
-            res["dataset"].name, enet_pred[:5], pred[:5])
-
-
-# TODO: More robust classification tests
-def assert_classification_result(results):
-    classification_results = [r for r in results if
-                              r["param"]["objective"] != "reg:squarederror"]
-    for res in classification_results:
-        # Check accuracy  is reasonable
-        assert res["eval"][-1] < 2.0, (res["dataset"].name, res["eval"][-1])
-
-
-class TestLinear(unittest.TestCase):
-
-    datasets = ["Boston", "Digits", "Cancer", "Sparse regression",
-                "Boston External Memory"]
-
-    @pytest.mark.skipif(**tm.no_sklearn())
-    def test_coordinate(self):
-        variable_param = {'booster': ['gblinear'], 'updater':
-                          ['coord_descent'], 'eta': [0.5], 'top_k':
-                          [10], 'tolerance': [1e-5], 'nthread': [2],
-                          'alpha': [.005, .1], 'lambda': [.005],
-                          'feature_selector': ['cyclic', 'shuffle',
-                                               'greedy', 'thrifty']}
-        for param in parameter_combinations(variable_param):
-            results = run_suite(param, 150, self.datasets, scale_features=True)
-            assert_regression_result(results, 1e-2)
-            assert_classification_result(results)
-
-    @pytest.mark.skipif(**tm.no_sklearn())
-    def test_shotgun(self):
-        variable_param = {'booster': ['gblinear'], 'updater':
-                          ['shotgun'], 'eta': [0.5], 'top_k': [10],
-                          'tolerance': [1e-5], 'nthread': [2],
-                          'alpha': [.005, .1], 'lambda': [.005],
-                          'feature_selector': ['cyclic', 'shuffle']}
-        for param in parameter_combinations(variable_param):
-            results = run_suite(param, 200, self.datasets, True)
-            assert_regression_result(results, 1e-2)
-            assert_classification_result(results)
+parameter_strategy = strategies.fixed_dictionaries({
+    'booster': strategies.just('gblinear'),
+    'eta': strategies.floats(0.01, 0.25),
+    'tolerance': strategies.floats(1e-5, 1e-2),
+    'nthread': strategies.integers(1, 4),
+    'alpha': strategies.floats(1e-5, 2.0),
+    'lambda': strategies.floats(1e-5, 2.0)
+})
+
+coord_strategy = strategies.fixed_dictionaries({
+    'feature_selector': strategies.sampled_from(['cyclic', 'shuffle',
+                                                 'greedy', 'thrifty']),
+    'top_k': strategies.integers(1, 10),
+})
+
+
+def train_result(param, dmat, num_rounds):
+    result = {}
+    xgb.train(param, dmat, num_rounds, [(dmat, 'train')], verbose_eval=False,
+              evals_result=result)
+    return result
+
+
+# Loss is not guaranteed to always decrease because of regularisation parameters
```

Comment:
Can we have a test that when regularization is not applied, the loss is guaranteed to decrease?

Reply:
Done